connections/stream-migration.md

```diff
@@ -0,0 +1,217 @@
+# Stream migration
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+|-----------------|---------------|--------|-----------------|
+| 1A              | Working Draft | Active | r0, 2022-04-13  |
+
+Authors: [@marcopolo]
+
+Interest Group: TODO
+## Introduction
+
+A peer may have many connections open for another peer and may be transmitting
+data on less optimal connections. For example a peer could be connected to
+another peer both directly and via a relay. In that case we'd like to move any
+streams from the relay over to the the better direct connection. A similar
+argument can be made with QUIC and TCP.
+
+This protocol attempts to solve the problem of how to seamlessly move a stream
+from one connection to another. This protocol also enables the peer to prune
+excess connections since they will no longer be used.
+
+## Requirements
+
+1. Transport agnostic. Really, this means migrating at the stream level.
+1. Minimal overhead. Overhead should be at most a small per-stream cost (no additional framing, etc.)
+1. No interruption. Reading/writing should be continuous.
+1. Transparent. Applications using migratable streams shouldn't notice anything.
+1. Correct. There can't be any ambiguity (one side believing the migration happened, the other side disagreeing, etc.).
+
+## The Protocol
+The goal of the protocol is to move traffic from one stream to another
+seamlessly. The final state of the new stream should be the same as the initial
+state of the old stream.
+
+![stream-migration](./stream-migration/stream-migration.svg)
+
+<details>
+  <summary>Instructions to reproduce diagram</summary>
+
+``` plantuml
+@startuml stream-migration
+skinparam sequenceMessageAlign center
+entity Initiator
+entity Responder
+
+note over Initiator, Responder: Assume at least 2 connections.
+
+Initiator <-> Responder: <b>Connection 1; stream A</b>
+Initiator <-> Responder: <b>Connection 2</b>
+
+== Stream Migration ==
+
+note over Initiator, Responder: Migrate <b>Stream A</b> to <b>Stream B</b>
+
+Initiator -> Responder: <b>Stream A:</b> Start stream migration, this stream is <b>id=A</b>
```

Comment:
This should probably be the other way around. The stream is named right after it is opened, and at some point later the migration is initiated.